gh-105802: email._header_value_parser: prevent IndexError in get_obs_local_part

fsc-eriker opened the pull request:
The parsing logic is supposed to raise an error if the header cannot be parsed, but I found a corner case where this didn't happen, and a different error ended up being raised, which caused Python to abort for some malformed messages.

Pardon any problems with this PR; this is my first one.

I have completed the CLA in 2021 (BPO user name eriker)

<!-- gh-issue-number: gh-105802 -->
* Issue: gh-105802
<!-- /gh-issue-number -->

fsc-eriker committed:
email._header_value_parser: prevent IndexError in get_obs_local_part

(GH-105802)

serhiy-storchaka commented on Lib/email/_header_value_parser.py:
```diff
@@ -1513,6 +1513,9 @@ def get_obs_local_part(value):
         obs_local_part.append(token)
+    if not obs_local_part:
+        raise(errors.HeaderParseError(
+            "abandoning parse; truncated value? ({})".format(value)))
```
It should not raise an exception, it should add it to the defects list.

fsc-eriker committed:
Update for GH-104802: register a defect instead of raising an error

Based on PR review

fsc-eriker committed:
Merge branch 'main' into fix-issue-gh-105802

fsc-eriker committed:
Fix formatting error in blurb

serhiy-storchaka commented on Lib/email/_header_value_parser.py:
```diff
@@ -1513,16 +1513,20 @@ def get_obs_local_part(value):
+    else:
+        if (obs_local_part[0].token_type == 'dot' or
             obs_local_part[0].token_type=='cfws' and
             obs_local_part[1].token_type=='dot'):
```
This can still fail if `len(obs_local_part) == 1 and obs_local_part[0].token_type=='cfws'`.

Can you add a corresponding test?

serhiy-storchaka commented on Lib/test/test_email/test__header_value_parser.py:
```diff
@@ -2588,6 +2588,22 @@ def test_get_msg_id_empty(self):
+        msg_id = self._test_get_x(
+            parser.get_msg_id,
+            '<[TeRriBlyLongBase64==@microsoft.com]>',
+            '<', # sic
+            '<', # sic
```
What do these "sic" mean?

fsc-eriker committed:
Tighten up obs_local_part checks further to avoid IndexError

Also, reformat slightly to make the indentation more obvious

fsc-eriker committed:
Merge branch 'main' into fix-issue-gh-105802

serhiy-storchaka commented on Lib/email/_header_value_parser.py:
```diff
@@ -1513,16 +1513,22 @@ def get_obs_local_part(value):
+        if (obs_local_part[0].token_type == 'dot'
+            or (obs_local_part[0].token_type=='cfws'
+                and len(obs_local_part) == 1
+                or obs_local_part[1].token_type=='dot')):
```
Please add tests for all these new cases. For example this condition is passed if there are more that 1 parts and the second part is a dot. Does it report a defect "Invalid leading '.' in local part" for "python.org"?

fsc-eriker committed:
Update for GH-104802: don't register dot defect if there is a single cwf

Don't add new error cases when refactoring; simply ensure that the code avoids an IndexError for the cases it attempts to handle
